File: src/45. Jump Game II.js
/**
 * @param {number[]} nums
 * @return {number}
 */

/*
function jump (nums) {
  // define a dp array which cell of index i represents min jumps from last index to index i
  const dp = Array(nums.length).fill(Infinity)
  dp[nums.length - 1] = 0
  for (let i = nums.length - 2; i >= 0; i--) {
    for (let j = nums[i]; j > 0; j--) {
      const jumpIndex = Math.min(i + j, nums.length - 1)
      dp[i] = Math.min(dp[jumpIndex] + 1, dp[i])
    }
  }
  return dp[0]
}
 */

export function jump(nums) {
  let jumps = 0,
    end = 0,
    farthest = 0
  for (let i = 0; i < nums.length - 1; i++) {
    farthest = Math.max(farthest, i + nums[i])
    if (i === end) {
      jumps++
      end = farthest
    }
  }
  return jumps
}

File: src/45. Jump Game II.cpp
// DP
// class Solution {
// public:
//     int jump(vector<int>& nums) {
//         int n = nums.size();
//         vector<int> dp(n, INT_MAX);
//         dp[n - 1] = 0;
//         for (int i = n - 2; i >= 0; --i) {
//             int furtherJump = min(i + nums[i], n - 1);
//             for (int j = i + 1; j <= furtherJump; ++j) {
//                 if (dp[j] < dp[i] - 1) dp[i] = dp[j] + 1;
//             }
//         }
//         return dp[0];
//     }
// };

// Greedy
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        int jumps = 0;
        int end = 0;
        int farthest = 0;
        for (int i = 0; i < n - 1; ++i) {
            farthest = max(farthest, i + nums[i]);
            if (i == end) {
                jumps++;
                end = farthest;
            }
        }
        return jumps;
    }
};